### PR TITLE
Feat/user management

### PR DESCRIPTION
**Pull request by AchmadQiz**
- add endpoint to get all user, get user by id, update user, update user status, delete user and sign up
- modify user model: add status and role column and modify password field length
- add error handling helper
- add email validation and hash password when sign up
- add DTO signup input and update input

### CODE REVIEW

File: app/main/controller/user_controller.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    register_user,
+    get_all_users,
+    get_a_user,
+    update_user,
+    update_user_status,
+    delete_user,
+)
+from ...extensions import ns
+
+@ns.route("/signup")
```
Comment (fatjan) [inline app/main/controller/user_controller.py:17]: Can we make this `/user/signup` ?

File: app/main/controller/user_controller.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+class UserList(Resource):
```
Comment (fatjan) [inline app/main/controller/user_controller.py:18]: I think we should rename this maybe `UserSignUp`?

File: app/main/controller/user_controller.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+class User(Resource):
+    def get(self, public_id):
+        """Get a user by its identifier"""
+        user = get_a_user(public_id)
+        return user
+
+    @ns.expect(_user, validate=True)
+    def put(self, public_id):
+        """Update a user"""
+        _updateduser = update_user(public_id, ns.payload)
```
Comment (fatjan) [inline app/main/controller/user_controller.py:41]: this should be renamed to `updated_user`

File: app/main/model/user.py [hunk 4]
```diff
@@ -1,38 +1,134 @@
...
+        
+    def delete_user(self, public_id):
+        try:
+            user = self.query.filter_by(public_id=public_id).first()
+            if not user:
+                raise Exception("User not found. Invalid ID")
+            
+            db.session.delete(user)  
+            db.session.commit()
+            return True
```
Comment (fatjan) [inline app/main/model/user.py:98]: why we return True here?

Comment (AchmadQiz) [reply app/main/model/user.py:98]: I think after deletion, we don't need to return anything, so I'm returning True as a sign if the deletion was successful. the response already in user_service

File: app/main/controller/user_controller.py (hunk 1 above)
Comment (AchmadQiz) [reply app/main/controller/user_controller.py]: i have pushed the revision

**Comment by fatjan** [comment]
> _ No description provided. _

Please add some description about your PR, thanks

File: app/main/controller/user_controller.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    def put(self, public_id):
+        """Update a user"""
+        _updateduser = updated_user(public_id, ns.payload)
+        return _updateduser
+
+    def delete(self, public_id):
+        """Delete a user"""
+        return delete_user(public_id)
+
+## work in progress, need authentication for this endpoint
```
Comment (fatjan) [inline app/main/controller/user_controller.py:48]: I think other endpoints also need authentication

File: app/main/service/user_service.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+            "status": "success",
+            "message": "Successfully update user",
+            "data": updated_user,
+        }
+        return response_object, 201
+    except Exception as e:
+        log.error(f"Error in update_user: {str(e)}")
+        return error_handler(e)
+
+def updated_user_status(public_id):
```
Comment (fatjan) [inline app/main/service/user_service.py:35]: this should be `update_user_status`

File: app/main/util/dto.py [hunk 7]
```diff
@@ -17,3 +17,13 @@ class ReviewDto:
             "region_id": fields.String(description="region Identifier"),
         },
     )
+
+class UserDto:
+    user = api.model(
+        "user",
+        {
+            "username": fields.String(required=True, description="username"),
```
Comment (fatjan) [inline app/main/util/dto.py:25]: should a user be able to login using username and also email? I mean one of them should be sufficient right?

Comment (AchmadQiz) [reply app/main/util/dto.py:25]: login endpoint only use email and password, using another dto. and this on feat/user_authentication branch

File: app/main/service/user_service.py (hunk 6 above)
Comment (AchmadQiz) [reply app/main/service/user_service.py]: done

File: app/main/controller/user_controller.py (hunk 5 above)
Comment (AchmadQiz) [reply app/main/controller/user_controller.py:49]: the authentication is on another branch, i think after this pr is merged

File: app/main/util/dto.py [hunk 8]
```diff
@@ -17,3 +17,19 @@ class ReviewDto:
...
+            "email": fields.String(required=True, description="user email"),
+            "password": fields.String(required=True, description="user password")
+        },
+    )
+    status = api.model(
+        "status",
+        {
+            "banned": fields.Boolean(required=True, description="username"),
```
Comment (fatjan) [inline app/main/util/dto.py:33]: why the field name is `banned` and the description is `username`?
I don't get it.

File: app/main/util/dto.py (hunk 7 above)
Comment (fatjan) [reply app/main/util/dto.py:25]: so, what's the point of having username?

Is it to be displayed on the UI?

File: tests/service/test_user_service.py [hunk 9]
```diff
@@ -0,0 +1,190 @@
+import uuid
+from unittest import TestCase
+from unittest.mock import patch
+from app import create_app
+from app.main.service.user_service import (
+    get_all_users,
+    get_a_user,
+    register_user,
+    delete_user,
+    updated_user,
```
Comment (fatjan) [inline tests/service/test_user_service.py:10]: please check your code again, I believe you updated these methods in the service. But you didn't update them in the test.

```
ImportError while importing test module '/Users/fatma/Documents/project-zusammen/hidden-gem/hidden-gem-backend/tests/service/test_user_service.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
../../../../.pyenv/versions/3.9.16/lib/python3.9/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
tests/service/test_user_service.py:5: in <module>
    from app.main.service.user_service import (
E   ImportError: cannot import name 'updated_user' from 'app.main.service.user_service' (/Users/fatma/Documents/project-zusammen/hidden-gem/hidden-gem-backend/app/main/service/user_service.py)
```

Comment (fatjan) [reply tests/service/test_user_service.py:10]: Please make sure to run your tests locally and ensure everything pass

**Review by fatjan — CHANGES_REQUESTED** [review]
Please update your tests and make sure everything pass

**Comment by AchmadQiz** [comment]
i have update the unittest, please kindly review

File: tests/controller/test_user_controller.py [hunk 10]
```diff
@@ -0,0 +1,181 @@
+# import json
```
Comment (fatjan) [inline tests/controller/test_user_controller.py:1]: why all these are commented?

**Review by fatjan — CHANGES_REQUESTED** [review]
Please update your test to be not being commented

File: tests/service/test_user_service.py [hunk 11]
```diff
@@ -0,0 +1,193 @@
+# import uuid
```
Comment (fatjan) [inline tests/service/test_user_service.py:1]: Please uncomment these tests

**Comment by AchmadQiz** [comment]
Sorry, It seems like I forgot to uncomment. I have pushed the update for uncomment.

**Review by fatjan — CHANGES_REQUESTED** [review]
Please resolve the conflicts, thank you

File: tests/service/test_user_service.py [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+        mock_update_user_status.return_value = data
+
+        # Act
+        response, status_code = update_user_status(public_id, payload)
+        result = response["data"]
+
+        # Assert
+        self.assertEqual(status_code, 201)
+        self.assertEqual(response["status"], "success")
+        self.assertEqual(response["message"], "Successfully update status user")
```
Comment (fatjan) [inline tests/service/test_user_service.py:174]: It should be `user status`

File: tests/model/test_user_model.py [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+        user_model = User()
+        user = user_model.register_user(user_data)
+
+        payload = {
+            "banned" : True
+        }
+
+        # ACT
+        updated_user = user_model.update_user_status(user["public_id"], payload)
+        print("\n\n\n\n\n")
```
Comment (fatjan) [inline tests/model/test_user_model.py:104]: please remove the print statements

**Review by fatjan — CHANGES_REQUESTED** [review]
please update as suggested

**Comment by AchmadQiz** [comment]
I have pushed the update

File: tests/controller/test_user_controller.py [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+    @patch("app.main.controller.user_controller.update_user_status")
+    def test_update_user_status(self, mock_update_user):
+        # ARRANGE
+        public_id = "test-public-id"
+        update_status_payload = {
+            "banned" : True,
+        }
+        expected_response = {
+            "status": "success",
+            "message": "Successfully update status user",
```
Comment (fatjan) [inline tests/controller/test_user_controller.py:167]: Don't you need to update this as well?

how did you run the test locally? did all pass?

Comment (fatjan) [reply tests/controller/test_user_controller.py:167]: How did you update your branch with the updated main branch?

I tried running the test in your branch locally. The test failed:

```
Traceback (most recent call last):
  File "/Users/fatma/.pyenv/versions/hgem/bin/flask", line 8, in <module>
    sys.exit(main())
  File "/Users/fatma/.pyenv/versions/3.9.16/envs/hgem/lib/python3.9/site-packages/flask/cli.py", line 1064, in main
    cli.main()
  File "/Users/fatma/.pyenv/versions/3.9.16/envs/hgem/lib/python3.9/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
  File "/Users/fatma/.pyenv/versions/3.9.16/envs/hgem/lib/python3.9/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/fatma/.pyenv/versions/3.9.16/envs/hgem/lib/python3.9/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/fatma/.pyenv/versions/3.9.16/envs/hgem/lib/python3.9/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
  File "/Users/fatma/.pyenv/versions/3.9.16/envs/hgem/lib/python3.9/site-packages/click/decorators.py", line 92, in new_func
    return ctx.invoke(f, obj, *args, **kwargs)
  File "/Users/fatma/.pyenv/versions/3.9.16/envs/hgem/lib/python3.9/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
  File "/Users/fatma/.pyenv/versions/3.9.16/envs/hgem/lib/python3.9/site-packages/flask/cli.py", line 912, in run_command
    raise e from None
  File "/Users/fatma/.pyenv/versions/3.9.16/envs/hgem/lib/python3.9/site-packages/flask/cli.py", line 898, in run_command
    app = info.load_app()
  File "/Users/fatma/.pyenv/versions/3.9.16/envs/hgem/lib/python3.9/site-packages/flask/cli.py", line 313, in load_app
    app = locate_app(import_name, None, raise_if_not_found=False)
  File "/Users/fatma/.pyenv/versions/3.9.16/envs/hgem/lib/python3.9/site-packages/flask/cli.py", line 219, in locate_app
    __import__(module_name)
  File "/Users/fatma/Documents/project-zusammen/hidden-gem/hidden-gem-backend/app/__init__.py", line 1, in <module>
    from .main import create_app
  File "/Users/fatma/Documents/project-zusammen/hidden-gem/hidden-gem-backend/app/main/__init__.py", line 14, in <module>
    from app.main.controller.review_controller import ns as review_ns
  File "/Users/fatma/Documents/project-zusammen/hidden-gem/hidden-gem-backend/app/main/controller/review_controller.py", line 5, in <module>
    _upvote = review_dto.upvote
AttributeError: 'ReviewDto' object has no attribute 'upvote'
```

Comment (fatjan) [reply tests/controller/test_user_controller.py:167]: It seems that the updated main branch from @hfzfdlh is not in your branch yet

**Comment by AchmadQiz** [comment]
I have pushed the update, bismillah

**Review by fatjan — APPROVED** [review]

**Pull request merged**